src/duckdb/write.rs

```diff
@@ -117,42 +119,71 @@ impl DataSink for DuckDBDataSink {
-        let duckdb_conn = DuckDB::duckdb_conn(&mut db_conn).map_err(to_datafusion_error)?;
-
-        let data_batches_result = data
-            .collect::<Vec<datafusion::common::Result<RecordBatch>>>()
-            .await;
-
-        let data_batches: Vec<RecordBatch> = data_batches_result
-            .into_iter()
-            .collect::<Result<Vec<_>, _>>()
-            .map_err(check_and_mark_retriable_error)?;
-
-        constraints::validate_batch_with_constraints(&data_batches, self.duckdb.constraints())
-            .await
-            .context(super::ConstraintViolationSnafu)
-            .map_err(to_datafusion_error)?;
-
-        let tx = duckdb_conn
-            .conn
-            .transaction()
-            .context(super::UnableToBeginTransactionSnafu)
-            .map_err(to_datafusion_error)?;
-
-        let num_rows = match **self.duckdb.constraints() {
-            [] => self.try_write_all_no_constraints(&tx, &data_batches)?,
-            _ => self.try_write_all_with_constraints(&tx, &data_batches)?,
-        };
+        let duckdb = Arc::clone(&self.duckdb);
+        let overwrite = self.overwrite;
+        let on_conflict = self.on_conflict.clone();
+
+        // Limit channel size to a maximum of 100 RecordBatches queued for cases when DuckDB is slower than the writer stream, 
+        // so that we don't significantly increase memory usage. After the maximum RecordBatches are queued, the writer stream will wait 
+        // until DuckDB is able to process more data.
+        let (batch_tx, batch_rx): (flume::Sender<RecordBatch>, flume::Receiver<RecordBatch>) =
+            flume::bounded(100);
+
+        let duckdb_write_handle: JoinHandle<datafusion::common::Result<u64>> =
+            tokio::task::spawn_blocking(move || {
+                let mut db_conn = duckdb.connect_sync().map_err(to_datafusion_error)?;
+
+                let duckdb_conn = DuckDB::duckdb_conn(&mut db_conn).map_err(to_datafusion_error)?;
+
+                let tx = duckdb_conn
+                    .conn
+                    .transaction()
+                    .context(super::UnableToBeginTransactionSnafu)
+                    .map_err(to_datafusion_error)?;
+
+                let num_rows = match **duckdb.constraints() {
+                    [] => try_write_all_no_constraints(duckdb, &tx, batch_rx.iter(), overwrite)?,
+                    _ => try_write_all_with_constraints(
+                        duckdb,
+                        &tx,
+                        batch_rx.iter(),
+                        overwrite,
+                        on_conflict,
+                    )?,
+                };
+
+                tx.commit()
+                    .context(super::UnableToCommitTransactionSnafu)
+                    .map_err(to_datafusion_error)?;
+
+                Ok(num_rows)
+            });
+
+        while let Some(batch) = data.next().await {
+            let batch = batch.map_err(check_and_mark_retriable_error)?;
```

Comment:
The error should also be passed to duckdb_write_handle task so we know about the error and don't commit/write the data.